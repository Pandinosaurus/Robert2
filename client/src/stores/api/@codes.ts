/* eslint-disable import/prefer-default-export */

/**
 * Codes d'erreur retournés par l'API.
 *
 * NOTE IMPORTANTE:
 * En cas de modif., pensez à aussi mettre à jour le fichier lié dans la partie back-end.
 * {@see {@link /server/src/App/Errors/Enums/ApiErrorCode.php}}
 */
export enum ApiErrorCode {
    /** Erreur inconnue ou sans prise en charge. */
    UNKNOWN = 0,

    //
    // - Erreurs liées aux formulaires.
    //

    /** Une erreur ou plusieurs erreurs de validation ce sont produites. */
    VALIDATION_FAILED = 400,

    /** Le payload fourni dans la requête ne doit pas être vide. */
    EMPTY_PAYLOAD = 401,
}
