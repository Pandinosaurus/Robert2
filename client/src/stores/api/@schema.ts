import { z } from '@/utils/validation';

import type { UnionToTupleString } from '@/utils/@types';
import type { AnyZodObject, ZodEnum, ZodTypeAny } from 'zod';

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const withPaginationEnvelope = <T extends ZodTypeAny>(dataSchema: T) => (
    z.object({
        data: dataSchema.array(),
        pagination: z.object({
            perPage: z.number().positive(),
            currentPage: z.number().nonnegative(),
            total: z.object({
                items: z.number().nonnegative(),
                pages: z.number().nonnegative(),
            }),
        }),
    })
);

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const withCountedEnvelope = <T extends ZodTypeAny>(dataSchema: T) => (
    z.object({
        data: dataSchema.array(),
        count: z.number().nonnegative(),
    })
);

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const withCsvImportResult = <T extends AnyZodObject>(mapping: T) => (
    z.strictObject({
        total: z.number().nonnegative(),
        success: z.number().nonnegative(),
        errors: z.array(z.strictObject({
            line: z.number().nonnegative(),
            message: z.string(),
            errors: z.array(z.strictObject({
                // - Requis pour patcher le `(T: AnyZodObject).keyof()` qui renvoi un `ZodEnum<never>` sinon.
                field: z.union([mapping.keyof() as any as ZodEnum<UnionToTupleString<keyof T['shape']>>, z.string()]),
                value: z.string().nullable(),
                error: z.string(),
            })),
        })),
    })
);
