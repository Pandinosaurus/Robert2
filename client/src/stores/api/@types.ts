//
// - Types liés à la pagination / tri.
//

/** Sens de tri. */
export enum Direction {
    /** Direction ascendante. */
    ASC = 'asc',

    /** Direction descendante. */
    DESC = 'desc',
}

export type SortableParams = {
    /** La colonne avec laquelle on veut trier le jeu de résultats. */
    orderBy?: string,

    /**
     * Le jeu de résultat doit-il être trié de manière ascendante selon la colonne choisie
     * ci-dessus (ou celle par défaut si aucun colonne n'a été explicitement choisie).
     *
     * - Si `1`, le jeu de résultat sera trié de manière ascendante.
     * - Si `0`, il sera trié de manière descendante.
     *
     */
    // TODO: Modifier ça pour quelque chose du genre : `{ direction: Direction }`.
    //       (il faudra adapter le component de tableau qui utilise cette notation obsolète).
    ascending?: 0 | 1 | boolean,
};

export type PaginationParams = {
    /**
     * La page dont on veut récupérer le résultat.
     *
     * @default 1
     */
    page?: number,

    /** Le nombre de résultats par page que l'on souhaite récupérer. */
    limit?: number,
};

export type ListingParams = (
    & {
        /**
         * Permet de rechercher un terme en particulier.
         *
         * @default undefined
         */
        search?: string,
    }
    & SortableParams
    & PaginationParams
);

//
// - Enveloppes.
//

export type PaginatedData<T> = {
    data: T,
    pagination: {
        perPage: number,
        currentPage: number,
        total: {
            items: number,
            pages: number,
        },
    },
};

export type CountedData<T> = {
    data: T,
    count: number,
};
