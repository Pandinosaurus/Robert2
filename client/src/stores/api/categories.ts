import { z } from '@/utils/validation';
import requester from '@/globals/requester';
import { SubCategorySchema } from './subcategories';

import type { SchemaInfer } from '@/utils/validation';

// ------------------------------------------------------
// -
// -    Schema / Enums
// -
// ------------------------------------------------------

export const CategorySchema = z.object({
    id: z.number(),
    name: z.string(),
});

export const CategoryDetailsSchema = CategorySchema.extend({
    sub_categories: z.lazy(() => SubCategorySchema.array()),
});

// ------------------------------------------------------
// -
// -    Types
// -
// ------------------------------------------------------

export type Category = SchemaInfer<typeof CategorySchema>;

export type CategoryDetails = SchemaInfer<typeof CategoryDetailsSchema>;

//
// - Edition
//

export type CategoryEdit = {
    name: string,
};

// ------------------------------------------------------
// -
// -    Fonctions
// -
// ------------------------------------------------------

const all = async (): Promise<CategoryDetails[]> => {
    const response = await requester.get('/categories');
    return CategoryDetailsSchema.array().parse(response.data);
};

const create = async (data: CategoryEdit): Promise<CategoryDetails> => {
    const response = await requester.post('/categories', data);
    return CategoryDetailsSchema.parse(response.data);
};

const update = async (id: Category['id'], data: Partial<CategoryEdit>): Promise<CategoryDetails> => {
    const response = await requester.put(`/categories/${id}`, data);
    return CategoryDetailsSchema.parse(response.data);
};

const remove = async (id: Category['id']): Promise<void> => {
    await requester.delete(`/categories/${id}`);
};

export default { all, create, update, remove };
