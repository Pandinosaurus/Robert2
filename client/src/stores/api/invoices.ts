import { z } from '@/utils/validation';

import type { SchemaInfer } from '@/utils/validation';

// ------------------------------------------------------
// -
// -    Schema / Enums
// -
// ------------------------------------------------------

const InvoiceTaxSchema = z.strictObject({
    name: z.string(),
    is_rate: z.boolean(),
    value: z.decimal(),
    total: z.decimal(),
});

export const InvoiceSchema = z.strictObject({
    id: z.number(),
    number: z.string(),
    date: z.datetime(),
    url: z.string(),
    total_without_taxes: z.decimal(),
    total_taxes: z.lazy(() => InvoiceTaxSchema.array()),
    total_with_taxes: z.decimal(),
    currency: z.currency(),
});

// ------------------------------------------------------
// -
// -    Types
// -
// ------------------------------------------------------

export type Invoice = SchemaInfer<typeof InvoiceSchema>;
