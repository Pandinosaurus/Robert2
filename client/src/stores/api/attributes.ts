import { z } from '@/utils/validation';
import requester from '@/globals/requester';
import { CategorySchema } from './categories';

import type { Category } from './categories';
import type { SchemaInfer } from '@/utils/validation';

// ------------------------------------------------------
// -
// -    Schema / Enums
// -
// ------------------------------------------------------

/** Entité concernée par l'attribut. */
export enum AttributeEntity {
    MATERIAL = 'material',
}

/** Type d'attribut. */
export enum AttributeType {
    STRING = 'string',
    TEXT = 'text',
    INTEGER = 'integer',
    FLOAT = 'float',
    BOOLEAN = 'boolean',
    DATE = 'date',
}

const AttributeBaseSchema = z.strictObject({
    id: z.number(),
    name: z.string(),
    entities: z.array(z.nativeEnum(AttributeEntity)),
});

// NOTE: Pour le moment, pas moyen de faire ça mieux en gardant l'objet `strict`.
// @see https://github.com/colinhacks/zod/discussions/3011#discussioncomment-7718731
export const AttributeSchema = z.discriminatedUnion('type', [
    AttributeBaseSchema.extend({
        type: z.literal(AttributeType.STRING),
        max_length: z.number().nullable(),
    }),
    AttributeBaseSchema.extend({
        type: z.enum([AttributeType.INTEGER, AttributeType.FLOAT]),
        unit: z.string().nullable(),
        is_totalisable: z.boolean().nullable().transform(
            (value: boolean | null) => value ?? false,
        ),
    }),
    AttributeBaseSchema.extend({
        type: z.literal(AttributeType.BOOLEAN),
    }),
    AttributeBaseSchema.extend({
        type: z.literal(AttributeType.DATE),
    }),
    AttributeBaseSchema.extend({
        type: z.literal(AttributeType.TEXT),
    }),
]);

// NOTE: Pour le moment, pas moyen de faire ça mieux en gardant l'objet `strict`.
// @see https://github.com/colinhacks/zod/discussions/3011#discussioncomment-7718731
export const AttributeWithValueSchema = z.discriminatedUnion('type', [
    AttributeBaseSchema.extend({
        type: z.literal(AttributeType.STRING),
        max_length: z.number().nullable(),
        value: z.string().nullable(),
    }),
    AttributeBaseSchema.extend({
        type: z.literal(AttributeType.TEXT),
        value: z.string().nullable(),
    }),
    AttributeBaseSchema.extend({
        type: z.enum([AttributeType.INTEGER, AttributeType.FLOAT]),
        unit: z.string().nullable(),
        is_totalisable: z.boolean().nullable().transform(
            (value: boolean | null) => value ?? false,
        ),
        value: z.number().nullable(),
    }),
    AttributeBaseSchema.extend({
        type: z.literal(AttributeType.BOOLEAN),
        value: z.boolean().nullable(),
    }),
    AttributeBaseSchema.extend({
        type: z.literal(AttributeType.DATE),
        value: z.day().nullable(),
    }),
]);

// NOTE: Pour le moment, pas moyen de faire ça mieux en gardant l'objet `strict`.
// @see https://github.com/colinhacks/zod/discussions/3011#discussioncomment-7718731
export const AttributeDetailsSchema = (() => {
    const baseSchema = AttributeBaseSchema.extend({
        categories: z.lazy(() => CategorySchema.array()),
    });

    return z.discriminatedUnion('type', [
        baseSchema.extend({
            type: z.literal(AttributeType.STRING),
            max_length: z.number().nullable(),
        }),
        baseSchema.extend({
            type: z.enum([AttributeType.INTEGER, AttributeType.FLOAT]),
            unit: z.string().nullable(),
            is_totalisable: z.boolean().nullable().transform(
                (value: boolean | null) => value ?? false,
            ),
        }),
        baseSchema.extend({
            type: z.enum([AttributeType.TEXT, AttributeType.BOOLEAN, AttributeType.DATE]),
        }),
    ]);
})();

// ------------------------------------------------------
// -
// -    Types
// -
// ------------------------------------------------------

export type Attribute = SchemaInfer<typeof AttributeSchema>;

export type AttributeWithValue = SchemaInfer<typeof AttributeWithValueSchema>;

export type AttributeDetails = SchemaInfer<typeof AttributeDetailsSchema>;

//
// - Edition
//

export type AttributeCreate = {
    name: string,
    entities: AttributeEntity[],
    type?: AttributeType,
    unit?: string,
    max_length?: string | number | null,
    is_totalisable?: boolean,
    categories: Array<Category['id']>,
};

export type AttributeEdit = Omit<AttributeCreate, 'type'>;

// ------------------------------------------------------
// -
// -    Fonctions
// -
// ------------------------------------------------------

const all = async (categoryId?: Category['id'] | 'none', entity?: AttributeEntity): Promise<AttributeDetails[]> => {
    const params = {
        ...(categoryId !== undefined ? { category: categoryId } : {}),
        ...(entity !== undefined) ? { entity } : {},
    };
    const response = await requester.get('/attributes', { params });
    return AttributeDetailsSchema.array().parse(response.data);
};

const one = async (id: Attribute['id']): Promise<AttributeDetails> => {
    const response = await requester.get(`/attributes/${id}`);
    return AttributeDetailsSchema.parse(response.data);
};

const create = async (data: AttributeCreate): Promise<AttributeDetails> => {
    const response = await requester.post('/attributes', data);
    return AttributeDetailsSchema.parse(response.data);
};

const update = async (id: Attribute['id'], data: AttributeEdit): Promise<AttributeDetails> => {
    const response = await requester.put(`/attributes/${id}`, data);
    return AttributeDetailsSchema.parse(response.data);
};

const remove = async (id: Attribute['id']): Promise<void> => {
    await requester.delete(`/attributes/${id}`);
};

export default { all, one, create, update, remove };
