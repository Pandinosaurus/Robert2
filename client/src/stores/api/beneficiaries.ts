import { z } from '@/utils/validation';
import requester from '@/globals/requester';
import { UserSchema } from './users';
import { CompanySchema } from './companies';
import { CountrySchema } from './countries';
import { EstimateSchema } from './estimates';
import { InvoiceSchema } from './invoices';
import { BookingExcerptSchema } from './bookings';
import {
    withPaginationEnvelope,
} from './@schema';

import type { SchemaInfer } from '@/utils/validation';
import type DateTime from '@/utils/datetime';
import type { Estimate } from './estimates';
import type { Invoice } from './invoices';
import type { BookingExcerpt } from './bookings';
import type {
    Direction,
    ListingParams,
    PaginatedData,
    PaginationParams,
} from './@types';

// ------------------------------------------------------
// -
// -    Schema / Enums
// -
// ------------------------------------------------------

export const BeneficiarySchema = z.strictObject({
    id: z.number(),
    user_id: z.number().nullable(),
    first_name: z.string(),
    last_name: z.string(),
    full_name: z.string(),
    reference: z.string().nullable(),
    // TODO [zod@>3.22.4]: Remettre `email()`.
    email: z.string().nullable(),
    phone: z.string().nullable(),
    company_id: z.number().nullable(),
    company: z.lazy(() => CompanySchema).nullable(),
    street: z.string().nullable(),
    postal_code: z.string().nullable(),
    locality: z.string().nullable(),
    country_id: z.number().nullable(),
    country: z.lazy(() => CountrySchema).nullable(),
    full_address: z.string().nullable(),
    note: z.string().nullable(),
});

export const BeneficiaryDetailsSchema = BeneficiarySchema.extend({
    user: z.lazy(() => UserSchema).nullable(),
    stats: z.strictObject({
        borrowings: z.number().nonnegative(),
    }),
});

// ------------------------------------------------------
// -
// -    Types
// -
// ------------------------------------------------------

export type Beneficiary = SchemaInfer<typeof BeneficiarySchema>;

export type BeneficiaryDetails = SchemaInfer<typeof BeneficiaryDetailsSchema>;

//
// - Edition
//

export type BeneficiaryEdit = {
    first_name: string,
    last_name: string,
    reference: string | null,
    email: string | null,
    phone: string | null,
    company_id: number | null,
    street: string | null,
    postal_code: string | null,
    locality: string | null,
    country_id: number | null,
    user_id?: number,
    pseudo?: string,
    password?: string,
    note: string | null,
};

//
// - Récupération
//

type GetAllParams = ListingParams & {
    /**
     * Permet de ne récupérer que les bénéficiaires dans la "corbeille".
     *
     * @default false
     */
    deleted?: boolean,
};

type GetBookingsParams = PaginationParams & {
    /**
     * Date à partir de laquelle on veut la liste des bookings.
     * Inclura aussi les bookings qui ont commencés avant mais se terminent après cette date.
     *
     * @default undefined
     */
    after?: DateTime,

    /**
     * Le sens dans lequel on veut récupérer les bookings :
     * - `Direction.ASC`: Du plus ancien au plus récent.
     * - `Direction.DESC`: Du plus récent au plus ancien.
     *
     * @default Direction.DESC
     */
    direction?: Direction,
};

// ------------------------------------------------------
// -
// -    Fonctions
// -
// ------------------------------------------------------

const all = async (params: GetAllParams = {}): Promise<PaginatedData<Beneficiary[]>> => {
    const response = await requester.get('/beneficiaries', { params });
    return withPaginationEnvelope(BeneficiarySchema).parse(response.data);
};

const one = async (id: Beneficiary['id']): Promise<BeneficiaryDetails> => {
    const response = await requester.get(`/beneficiaries/${id}`);
    return BeneficiaryDetailsSchema.parse(response.data);
};

const create = async (data: BeneficiaryEdit): Promise<BeneficiaryDetails> => {
    const response = await requester.post('/beneficiaries', data);
    return BeneficiaryDetailsSchema.parse(response.data);
};

const update = async (id: Beneficiary['id'], data: BeneficiaryEdit): Promise<BeneficiaryDetails> => {
    const response = await requester.put(`/beneficiaries/${id}`, data);
    return BeneficiaryDetailsSchema.parse(response.data);
};

const restore = async (id: Beneficiary['id']): Promise<BeneficiaryDetails> => {
    const response = await requester.put(`/beneficiaries/restore/${id}`);
    return BeneficiaryDetailsSchema.parse(response.data);
};

const remove = async (id: Beneficiary['id']): Promise<void> => {
    await requester.delete(`/beneficiaries/${id}`);
};

const bookings = async (id: Beneficiary['id'], params: GetBookingsParams = {}): Promise<PaginatedData<BookingExcerpt[]>> => {
    const response = await requester.get(`/beneficiaries/${id}/bookings`, { params });
    return withPaginationEnvelope(BookingExcerptSchema).parse(response.data);
};

const estimates = async (id: Beneficiary['id']): Promise<Estimate[]> => {
    const response = await requester.get(`/beneficiaries/${id}/estimates`);
    return EstimateSchema.array().parse(response.data);
};

const invoices = async (id: Beneficiary['id']): Promise<Invoice[]> => {
    const response = await requester.get(`/beneficiaries/${id}/invoices`);
    return InvoiceSchema.array().parse(response.data);
};

export default {
    all,
    one,
    create,
    update,
    restore,
    remove,
    bookings,
    estimates,
    invoices,
};
