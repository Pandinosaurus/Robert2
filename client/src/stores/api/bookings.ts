import { z } from '@/utils/validation';
import requester from '@/globals/requester';
import { BeneficiarySchema } from './beneficiaries';
import { withPaginationEnvelope } from './@schema';
import {
    EventExtraSchema,
    EventTechnicianSchema,
    createEventDetailsSchema,
    EventMaterialBillableSchema,
    EventMaterialNotBillableSchema,
} from './events';

import type Decimal from 'decimal.js';
import type Period from '@/utils/period';
import type { ZodRawShape } from 'zod';
import type { Tax } from '@/stores/api/taxes';
import type { Material, UNCATEGORIZED } from '@/stores/api/materials';
import type { Category } from '@/stores/api/categories';
import type { Park } from '@/stores/api/parks';
import type { SchemaInfer } from '@/utils/validation';
import type {
    EventTax,
    EventTaxTotal,
} from '@/stores/api/events';
import type {
    PaginatedData,
    SortableParams,
    PaginationParams,
} from './@types';

// ------------------------------------------------------
// -
// -    Schema / Enums
// -
// ------------------------------------------------------

export enum BookingEntity {
    EVENT = 'event',
}

//
// - Schemas secondaires
//

const BookingMaterialNotBillableSchema = EventMaterialNotBillableSchema;
const BookingMaterialBillableSchema = EventMaterialBillableSchema;
const BookingMaterialSchema = z.union([
    BookingMaterialNotBillableSchema,
    BookingMaterialBillableSchema,
]);

const BookingExtraSchema = EventExtraSchema;

//
// - Schemas principaux
//

// - Booking excerpt schema.
export const BookingExcerptSchema = (() => z.strictObject({
    id: z.number(),
    entity: z.literal(BookingEntity.EVENT),
    title: z.string(),
    location: z.string().nullable(),
    color: z.string().nullable(),
    mobilization_period: z.period(),
    operation_period: z.period(),
    beneficiaries: z.lazy(() => BeneficiarySchema.array()),
    technicians: z.lazy(() => EventTechnicianSchema.array()),
    is_confirmed: z.boolean(),
    is_archived: z.boolean(),
    is_departure_inventory_done: z.boolean(),
    is_return_inventory_done: z.boolean(),
    has_not_returned_materials: z.boolean().nullable(),
    categories: z.number().array(), // - Ids des catégories liés.
    parks: z.number().array(), // - Ids des parcs liés.
    created_at: z.datetime(),
}))();

// - Booking summary schema.
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const createBookingSummarySchema = <T extends ZodRawShape>(augmentation: T) => z.strictObject({
    id: z.number(),
    entity: z.literal(BookingEntity.EVENT),
    title: z.string(),
    reference: z.string().nullable(),
    description: z.string().nullable(),
    location: z.string().nullable(),
    color: z.string().nullable(),
    mobilization_period: z.period(),
    operation_period: z.period(),
    beneficiaries: z.lazy(() => BeneficiarySchema.array()),
    technicians: z.lazy(() => EventTechnicianSchema.array()),
    materials_count: z.number().nonnegative(),
    is_confirmed: z.boolean(),
    is_billable: z.boolean(),
    is_archived: z.boolean(),
    is_departure_inventory_done: z.boolean(),
    is_return_inventory_done: z.boolean(),
    has_missing_materials: z.boolean().nullable(),
    has_not_returned_materials: z.boolean().nullable(),
    categories: z.number().array(), // - Ids des catégories liés.
    parks: z.number().array(), // - Ids des parcs liés.
    created_at: z.datetime(),
}).extend<T>(augmentation);

export const BookingSummarySchema = createBookingSummarySchema({});

// - Booking schema.
// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
export const createBookingSchema = <T extends ZodRawShape>(augmentation: T) => z.lazy(() => (
    createEventDetailsSchema({
        entity: z.literal(BookingEntity.EVENT),
        ...augmentation,
    })
));
export const BookingSchema = createBookingSchema({});

// ------------------------------------------------------
// -
// -    Types
// -
// ------------------------------------------------------

//
// - Main Types
//

type NarrowBooking<Schema, Entity extends BookingEntity> = (
    Extract<Schema, { entity: Entity }>
);

export type BookingExcerpt<Entity extends BookingEntity = BookingEntity> = (
    NarrowBooking<SchemaInfer<typeof BookingExcerptSchema>, Entity>
);

export type BookingSummary<Entity extends BookingEntity = BookingEntity> = (
    NarrowBooking<SchemaInfer<typeof BookingSummarySchema>, Entity>
);

export type Booking<IsBillable extends boolean = boolean, Entity extends BookingEntity = BookingEntity> = (
    IsBillable extends true
        ? Extract<SchemaInfer<typeof BookingSchema>, { entity: Entity, is_billable: true }>
        : Extract<SchemaInfer<typeof BookingSchema>, { entity: Entity, is_billable: false }>
);

//
// - Secondary Types
//

export type BookingMaterial<IsBillable extends boolean = boolean> =
    IsBillable extends true
        ? SchemaInfer<typeof BookingMaterialBillableSchema>
        : SchemaInfer<typeof BookingMaterialNotBillableSchema>;

export type BookingExtra = SchemaInfer<typeof BookingExtraSchema>;

export type BookingTax = EventTax;
export type BookingTaxTotal = EventTaxTotal;

//
// - Édition
//

export type MaterialQuantity = {
    id: Material['id'],
    quantity: number,
};

export type MaterialBillingData = {
    id: Material['id'],
    unit_price: Decimal,
    discount_rate: Decimal,
};

export type ExtraBillingTaxData = {
    name: string,
    is_rate: boolean,
    value: Decimal,
};

export type ExtraBillingData = {
    id: number | null,
    description: string | null,
    quantity: number,
    unit_price: Decimal | null,
    tax_id: Tax['id'] | null,
    taxes?: ExtraBillingTaxData[],
};

export type BillingData = {
    materials: MaterialBillingData[],
    extras: ExtraBillingData[],
    global_discount_rate: Decimal,
};

export type MaterialResynchronizableField =
    | 'name'
    | 'reference'
    | 'unit_price'
    | 'degressive_rate'
    | 'taxes';

export type ExtraResynchronizableField =
    | 'taxes';

//
// - Récupération
//

export type BookingListFilters = {
    period?: Period,
    search?: string,
    category?: Category['id'] | typeof UNCATEGORIZED,
    park?: Park['id'],
    endingToday?: boolean,
    returnInventoryTodo?: boolean,
    archived?: boolean,
    notConfirmed?: boolean,
};

type GetAllParamsPaginated = BookingListFilters & SortableParams & PaginationParams & { paginated?: true };
type GetAllInPeriodParams = { paginated: false, period: Period };

// ------------------------------------------------------
// -
// -    Fonctions
// -
// ------------------------------------------------------

async function all(params: GetAllInPeriodParams): Promise<BookingExcerpt[]>;
async function all(params?: GetAllParamsPaginated): Promise<PaginatedData<BookingExcerpt[]>>;
async function all({ period, ...params }: GetAllParamsPaginated | GetAllInPeriodParams = {}): Promise<unknown> {
    const normalizedParams = { paginated: true, ...params, ...period?.toQueryParams('period') };
    const response = await requester.get('/bookings', { params: normalizedParams });

    return normalizedParams.paginated
        ? withPaginationEnvelope(BookingExcerptSchema).parse(response.data)
        : BookingExcerptSchema.array().parse(response.data);
}

const oneSummary = async (entity: BookingEntity, id: Booking['id']): Promise<BookingSummary> => {
    const response = await requester.get(`/bookings/${entity}/${id}/summary`);
    return BookingSummarySchema.parse(response.data);
};

const one = async (entity: BookingEntity, id: Booking['id']): Promise<Booking> => {
    const response = await requester.get(`/bookings/${entity}/${id}`);
    return BookingSchema.parse(response.data);
};

const updateMaterials = async (entity: BookingEntity, id: Booking['id'], materials: MaterialQuantity[]): Promise<Booking> => {
    const response = await requester.put(`/bookings/${entity}/${id}/materials`, materials);
    return BookingSchema.parse(response.data);
};

const updateBilling = async (entity: BookingEntity, id: Booking['id'], data: BillingData): Promise<Booking> => {
    const response = await requester.put(`/bookings/${entity}/${id}/billing`, data);
    return BookingSchema.parse(response.data);
};

const resynchronizeMaterial = async (
    entity: BookingEntity,
    id: Booking['id'],
    materialId: BookingMaterial['id'],
    selection: MaterialResynchronizableField[],
): Promise<BookingMaterial> => {
    const response = await requester.put(`/bookings/${entity}/${id}/materials/${materialId}/resynchronize`, selection);
    return BookingMaterialSchema.parse(response.data);
};

const resynchronizeExtra = async (
    entity: BookingEntity,
    id: Booking['id'],
    extraId: BookingExtra['id'],
    selection: ExtraResynchronizableField[],
): Promise<BookingExtra> => {
    const response = await requester.put(`/bookings/${entity}/${id}/extras/${extraId}/resynchronize`, selection);
    return BookingExtraSchema.parse(response.data);
};

export default {
    all,
    one,
    oneSummary,
    updateMaterials,
    updateBilling,
    resynchronizeMaterial,
    resynchronizeExtra,
};
