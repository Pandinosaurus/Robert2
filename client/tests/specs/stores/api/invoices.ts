import { InvoiceSchema } from '@/stores/api/invoices';
import data from '@fixtures/invoices';

import type { SafeParseSuccess } from 'zod';

describe('Invoices Api', () => {
    test('Schema', () => {
        data.default().forEach((datum: any) => {
            const result = InvoiceSchema.safeParse(datum);
            expect(result.success).toBeTruthy();
            expect((result as SafeParseSuccess<unknown>).data).toMatchSnapshot();
        });
    });
});
