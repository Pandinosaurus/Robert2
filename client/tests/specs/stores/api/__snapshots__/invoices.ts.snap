// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invoices Api Schema 1`] = `
{
  "currency": /* Currency Instance => */ "EUR",
  "date": /* DateTime Instance => */ "2020-01-30 14:00:00",
  "id": 1,
  "number": "2020-00001",
  "total_taxes": [
    {
      "is_rate": true,
      "name": "T.V.A.",
      "total": /* Decimal Instance => */ "108.82",
      "value": /* Decimal Instance => */ "20",
    },
  ],
  "total_with_taxes": /* Decimal Instance => */ "652.96",
  "total_without_taxes": /* Decimal Instance => */ "544.13",
  "url": "http://loxya.test/invoices/1/pdf",
}
`;
